srml/staking/src/lib.rs

```diff
@@ -514,50 +711,19 @@ impl<T: Trait> Module<T> {
 			});
 		}
 
-		let event = if new_slash_count > grace {
-			let slash = {
-				let base_slash = Self::current_offline_slash();
-				let instances = slash_count - grace;
-
-				let mut total_slash = BalanceOf::<T>::default();
-				for i in instances..(instances + count as u32) {
-					if let Some(total) = base_slash.checked_shl(i)
-							.and_then(|slash| total_slash.checked_add(&slash)) {
-						total_slash = total;
-					} else {
-						// reset slash count only up to the current
-						// instance. the total slash overflows the unit for
-						// balance in the system therefore we can slash all
-						// the slashable balance for the account
-						<SlashCount<T>>::insert(v.clone(), slash_count + i);
-						total_slash = Self::slashable_balance(&v);
-						break;
-					}
-				}
-
-				total_slash
-			};
+		let prefs = Self::validators(&v);
+		let unstake_threshold = prefs.unstake_threshold.min(MAX_UNSTAKE_THRESHOLD);
+		let max_slashes = grace + unstake_threshold;
 
+		let event = if new_slash_count > max_slashes {
+			// They're bailing.
+			let slash = Self::current_offline_slash()
+				.checked_shl(prefs.unstake_threshold)
+				.unwrap_or_else(Self::slot_stake);
```

Comment:
Unclear to me what `slot_stake` is. Comment above the storage value says:

>		/// The highest and lowest staked validator slashable balances.

But it is only a single value...

Elsewhere in the code comments it says:

```
		// Figure out the minimum stake behind a slot.
		let slot_stake = candidates.last().map(|i| i.1.total).unwrap_or_default();
		<SlotStake<T>>::put(&slot_stake);
```

Which implies it is a minimum. But then it does not make sense to be used here. I am thinking the logic should be the following:

* Either punish the user by `current_offline_slash` * exponential weighting factor

and if that overflows

* Punish the user by the max stake of any validator (which is then limited by how much they actually exposed in `slash_validator`

So a `max` value makes sense to me...

Reply:
Updated the docs. It's meant to be the amount of the least staked active validator.

Reply:
@gavofyork Does this imply that all validators are only held accountable to the amount of stake staked by the minimum validator?

i.e. if a validator gets chosen with some small amount of stake, that all other validators can only be punished for being offline by that small amount? (hence the fallback)

But if that is the case, why do we only fallback to this when the exponential weighting factor overflows? Doesn't this cause a situation where users will choose high `unstake_threshold` such that they get a lesser penalty for doing "more bad"

Reply:
I agree with the argument @shawntabrizi mentioned, in a nutshell, one validator (given that it has the funds) can choose a `stash_balance` + `unstake_threshold` combination in a way the `.checked_shl(unstake_threshold).unwrap_or_else(Self::slot_stake);` overflows and then benefit from being slashed only as if it had `Self::slot_stake` in stash => less punishment than what it deserves. 

The counter-argument could be that with the typically expected balances that validators will put in the stash and the fact that max unstake_threshold is 10 this is not really feasible to happen?

Reply:
as explained in chat, `slot_stake` is the amount by which all slots are staked. the security parameters of the system assume that each validator slot (and the stakers behind them) may be punished by at least this amount and thus it should be used as the maximum. with `MAX_UNSTAKE_THRESHOLD` and `OfflineSlash` that are too high then theoretically a validator could have a theoretical punishment that is greater than `slot_stake` but not so big as to overflow the `Balance`. in this case, it would be a little unfair as they would be better to aim to overflow. however, it's not expected that the parameters would ever let this situation happen, and even if it did, it would be such a large write off anyway, it's not like it's compromising the system's security in any way. it's potentially going to slash a little more than it *really* should.

as discussed elsewhere, i'm happy to cap it at `slot_stake` anyway.